source/server/drain_manager_impl.cc

```diff
@@ -14,7 +16,34 @@ namespace Server {
 
 DrainManagerImpl::DrainManagerImpl(Instance& server,
                                    envoy::config::listener::v3::Listener::DrainType drain_type)
-    : server_(server), drain_type_(drain_type) {}
+    : server_(server), drain_type_(drain_type), cbs_(server_.dispatcher()) {}
+
+DrainManagerImpl::DrainManagerImpl(Instance& server,
+                                   envoy::config::listener::v3::Listener::DrainType drain_type,
+                                   Event::Dispatcher& dispatcher)
+    : server_(server), drain_type_(drain_type), cbs_(dispatcher) {}
+
+DrainManagerSharedPtr
+DrainManagerImpl::createChildManager(Event::Dispatcher& dispatcher,
+                                     envoy::config::listener::v3::Listener::DrainType drain_type) {
+  auto child = std::make_shared<DrainManagerImpl>(server_, drain_type, dispatcher);
+
+  // Wire up the child so that when the parent starts draining, the child also sees the state-change
+  child_drain_cbs_.push_back(addOnDrainCloseCb(
+      dispatcher,
+      [child_ = std::weak_ptr<DrainManagerImpl>(child)](std::chrono::milliseconds) mutable {
+        // disregard the specified delay time and kick off the child drain immediately
```

Comment:
It seems that we're not spreading events when processing in child managers.  This will result in bursts of event callbacks when the child manager is notified.

Reply:
Correct that child managers are notified immediately, but the same delay logic will be applied according to each child's configuration in the following `startDrainSequence` call. Essentially the draining happens at the leaf's of the tree and I would not expect an exceedingly large number of DrainManagers in the tree.